### ISSUE TITLE
Enabling Node-to-Node Mesh After Route Reflector Configuration Can Stop Bird.cfg Update

### ISSUE THREAD

**Issue by bozoli** [projectcalico/calico#3785]
After enabling Route Reflector in Calico, if nodeToNodeMeshEnabled is re-enabled by mistake the node-to-node mesh peers stay in "start" state and from the moment nodeToNodeMeshEnabled is disabled again Calico stops updating the Bird configuration files, no matter what change is made.

## Expected Behavior

When enabling and disabling node-to-node mesh again with route reflector configured, everything should return to the same configuration and behavior.

## Current Behavior

When I enable and disable node-to-node mesh after route reflector is configured, Calico stops updating bird.cfg file, no matter what's the configuration change.

## Steps to Reproduce (for bugs)

1. Set nodeToNodeMeshEnabled to false
2. Configure one node as route reflector and the other nodes to peer with it
3. Set nodeToNodeMeshEnabled to true
4. Set nodeToNodeMeshEnabled to false

### Commands executed:
```
calicoctl apply -f - <<EOF
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  asNumber: 65100
EOF

calicoctl patch bgpconfiguration default -p '{"spec": {"nodeToNodeMeshEnabled": false}}'
calicoctl patch node <1st node name> -p '{"spec": {"bgp": {"routeReflectorClusterID": "244.0.0.1"}}}'

kubectl label node <1st node name> route-reflector=true

calicoctl apply -f - <<EOF
kind: BGPPeer
apiVersion: projectcalico.org/v3
metadata:
  name: peer-with-route-reflectors
spec:
  nodeSelector: all()
  peerSelector: route-reflector == 'true'
EOF

calicoctl apply -f - <<EOF
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  serviceClusterIPs:
  - cidr: <Service CIDR>
  asNumber: 65100
EOF

calicoctl patch bgpconfiguration default -p '{"spec": {"nodeToNodeMeshEnabled": false}}'
```

Status after the commands:
```
IPv4 BGP status
+----------------+-------------------+-------+----------+-------------+
|  PEER ADDRESS  |     PEER TYPE     | STATE |  SINCE   |    INFO     |
+----------------+-------------------+-------+----------+-------------+
| xxx.xxx.184.82 | node specific     | up    | 12:50:09 | Established |
| xxx.xxx.184.83 | node specific     | up    | 12:50:09 | Established |
| xxx.xxx.184.82 | node-to-node mesh | start | 12:50:24 | Idle        |
| xxx.xxx.184.83 | node-to-node mesh | start | 12:50:24 | Idle        |
+----------------+-------------------+-------+----------+-------------+
```
Extract of Bird.cfg from one of the nodes after the commands:
```
# For peer /host/<1st node name>/ip_addr_v4
protocol bgp Mesh_xxx_xxx_184_81 from bgp_template {
  neighbor xxx.xxx.184.81 as 65100;
}

# For peer /host/<2nd node name>/ip_addr_v4
# Skipping ourselves (xxx.xxx.184.82)

# For peer /host/<3rd node name>/ip_addr_v4
protocol bgp Mesh_xxx_xxx_184_83 from bgp_template {
  neighbor xxx.xxx.184.83 as 65100;
  passive on; # Mesh is unidirectional, peer will connect to us.
}
```

## Context

While I was running some tests with Calico, we were testing with different route reflector configuration, and with full-mesh configuration and I was hit by this issue several times. The only way I could find to get Calico back was destroying the cluster, remove manually all routes created by Bird, and redeploy the cluster (the cluster was failing to be redeployed without removing the routes created by Bird).

## Your Environment

* Calico version 3.15
* Orchestrator version (e.g. kubernetes, mesos, rkt): Rancher Kubernetes Engine 1.18.3
* Operating System and version: Ubuntu 18.04 LTS

**Comment by bozoli** [comment]
In fact the problem is happening in other configurations without RR as well. So I'm closing this issue.

**Comment by bozoli** [comment]
I found out that this issue only happens when I create the cluster setting CALICO_IPV4POOL_BLOCK_SIZE to /24 instead of leaving the default /26.

**Comment by bozoli** [comment]
In fact if I get to the state that the output of `calicoctl node status` is similar to the one below, Bird.cfg stops been updated. In any configuration I do. 
In my experience this usually happens when Route Reflector is configured and later the configuration is removed.

```
IPv4 BGP status
+-----------------+-------------------+-------+----------+-------------+
|  PEER ADDRESS   |     PEER TYPE     | STATE |  SINCE   |    INFO     |
+-----------------+-------------------+-------+----------+-------------+
| xxx.xxx.184.82  | node specific     | up    | 14:23:37 | Established |
| xxx.xxx.184.83  | node specific     | up    | 14:23:40 | Established |
| xxx.xxx.184.82  | node-to-node mesh | start | 15:03:04 | Idle        |
| xxx.xxx.184.83  | node-to-node mesh | start | 15:03:04 | Idle        |
| xxx.xxx.181.254 | global            | up    | 15:03:05 | Established |
| xxx.xxx.183.254 | global            | up    | 15:03:05 | Established |
+-----------------+-------------------+-------+----------+-------------+
```

**Comment by bozoli** [comment]
You can find below the Debug level logs of one of the Calico nodes. I executed the configuration commands around 15:37
[Calico-Node Logs.txt](https://github.com/projectcalico/calico/files/4920304/Calico-Node.Logs.txt)

**Comment by bozoli** [comment]
Here I crashed Calico configuration again, but this time I got the each executed command with the time and the logs.

[Executed Commands with Time.txt](https://github.com/projectcalico/calico/files/4920449/Executed.Commands.with.Time.txt)
[Calico-Node Logs2.txt](https://github.com/projectcalico/calico/files/4920450/Calico-Node.Logs2.txt)

**Comment by bozoli** [comment]
I found out that the problem isn't happening only with RR configurations, but also with normal configurations when I configure the  serviceClusterIPs or a serviceExternalIPs to be announced. After that all the commands I execute don't change anything in bird.cfg.

I replaced the images on the calico.yaml file by the images provided by Rancher (the ones below), using Calico version 3.13.4, and I had no issues so far, even with RR config (but I'll test better tomorrow). I don't know if the images have any customization or if they are just older images of Calico.
rancher/calico-node:v3.13.4
rancher/calico-cni:v3.13.4
rancher/calico-kube-controllers:v3.13.4
rancher/calico-ctl:v3.13.4
rancher/calico-pod2daemon-flexvol:v3.13.4

**Comment by bozoli** [comment]
Confirming that Calico v3.13.4 is working fine for me (not using Calico provided images).
I hope to test v3.15.1 soon.

**Comment by bozoli** [comment]
Tested with v3.15.1 and everything is working fine.
Tested again with v3.15.0 and the issue still there. 

The problem doesn't seems to be the RR configuration only as I said in the ticket, as I can make the Bird.cfg stop being updated with the following command just after creating the cluster:

```

calicoctl apply -f - <<EOF
apiVersion: projectcalico.org/v3
kind: BGPConfiguration
metadata:
  name: default
spec:
  logSeverityScreen: Debug
  asNumber: 65100
  serviceClusterIPs:
  - cidr: 10.46.32.0/24
  serviceExternalIPs:
  - cidr: 10.46.33.0/24
EOF
```

After executing this all other commands I execute to change the BGP configuration don't affect bird.cfg when running Calico v3.15.0.

As this seems to be a fixed issue on the last version, should this issue be closed?

**Comment by bozoli** [comment]
In reality the issue wasn't the RR configuration but the service IP advertisement that I was executing before changing some RR configuration. This issue was fixed in v3.15.1 (https://github.com/projectcalico/confd/pull/337).
So I'm closing this issue.
